File: y2023/day1/Day1.kt
package y2023.day1

import java.io.File

class Day1() {
    var lines = File("y2023/day1/day1.txt").useLines { it.toList() }
    val numset = mapOf(
        "1" to 1,
        "2" to 2,
        "3" to 3,
        "4" to 4,
        "5" to 5,
        "6" to 6,
        "7" to 7,
        "8" to 8,
        "9" to 9,
        "one" to 1,
        "two" to 2,
        "three" to 3,
        "four" to 4,
        "five" to 5,
        "six" to 6,
        "seven" to 7,
        "eight" to 8,
        "nine" to 9,
    )

    fun part1() {
        var res = 0
        for (ln in lines) {
            for(c in ln.toCharArray()) {
                if (c in '0'..'9') {
                    res += 10 * (c - '0')
                    break
                }
            }
            for(c in ln.toCharArray().reversed()) {
                if (c in '0'..'9') {
                    res += c - '0'
                    break
                }
            }
        }
        println(res)
    }

    fun part2() {
        var res = 0
        for (ln in lines) {
            var (first, last) = arrayOf(0, 0)
            var (firstIdx, lastIdx) = arrayOf(ln.length, -1)
            for (entry in numset) {
                val (s, v) = entry
                val idx = ln.indexOf(s)
                if (idx in 0..<firstIdx) {
                    firstIdx = idx
                    first = v * 10
                }
            }
        }
        println(res)
    }
}

fun main() {

}

File: y2023/day1/day1.go
package day1

import (
	"adv_of_code/helper"
	"os"
	"strings"
)

func Part1() {
	ct, err := os.ReadFile("y2023/day1/day1.txt")
	helper.Check(err)
	lines := strings.Split(string(ct), "\n")
	res := 0
	for _, ln := range lines {
		for _, c := range ln {
			if '0' <= c && c <= '9' {
				res += 10 * int(c-'0')
				break
			}
		}
		for i := len(ln) - 1; i >= 0; i-- {
			c := ln[i]
			if '0' <= c && c <= '9' {
				res += int(c - '0')
				break
			}
		}
	}
	print(res)
}

var numMap = map[string]int{
	"1":     1,
	"2":     2,
	"3":     3,
	"4":     4,
	"5":     5,
	"6":     6,
	"7":     7,
	"8":     8,
	"9":     9,
	"one":   1,
	"two":   2,
	"three": 3,
	"four":  4,
	"five":  5,
	"six":   6,
	"seven": 7,
	"eight": 8,
	"nine":  9,
}

func Part2() {
	ct, err := os.ReadFile("2023/day1/day1.txt")
	helper.Check(err)
	lines := strings.Split(string(ct), "\n")
	res := 0
	for _, ln := range lines {
		first, last := 0, 0
		fIdx, lIdx := len(ln)+1, -1
		for s, v := range numMap {
			idx := strings.Index(ln, s)
			if idx >= 0 && idx < fIdx {
				fIdx = idx
				first = v * 10
			}
			idx = strings.LastIndex(ln, s)
			if idx >= lIdx {
				lIdx = idx
				last = v
			}
		}
		res = first + last
	}
	print(res)
}

var Export = helper.Export{
	Part1: Part1,
	Part2: Part2,
}
